psp-uri.md

```diff
@@ -0,0 +1,102 @@
+
+A resource can be:
+
+- A blockchain. e.g. Polkadot, Ethereum
+- A runtime module in a chain. e.g. Balances in Polkadot (DOT), Crowdfunding module in Polkadot
+- A resource provided by a runtime module in a chain. e.g. An instance of a smart contract in a contracts module in a smart contract chain
+
+More examples:
+
+- A native fungible token of a chain. e.g. DOT on Polkadot, BTC on Bitcoin, ETH on Ethereum
+- A non-fungible token provided by a chain.
+- An instance of a non-fungible token. e.g. A particular Crypto Kitty
+- A bridged token provided by a chain. e.g. XBTC on ChainX
+- A utility token provided by runtime modules or smart contracts. e.g. MKR on Ethereum, SDOT on ChainX
+- A service provided by a chain. e.g. DEX on exchange chain
+- A trading pair in a exchange that provided by a chain. e.g. BTCETH pair on a DEX chain
+
+## Specification
+
+### URI format
+
+A URI is made of various parts:
+
+```
+URI = scheme:[//authority]path[?query][#fragment]
+authority = [userinfo@]host[:port]
+```
+
+For SURI, the format is defined as
+
+- schema: TBD
+  - `polkadot` or `substrate` or something more generalized?
+- port: N/A
+- userinfo: the parachain ID
+- host: the relay chain index or address
+  - Leave empty to point to current chain
+- path: the resource path in a chain
+  - Begin with the runtime moudle name and followed by runtime module specific path
+- query & fragment: unspecified
+
+### Encoding
+
+There are two encoding formats for SURI.
+
+- Standard encoding
+  - Encoded as a URI string
+
+- Custom SCALE encoding to avoid unnecessary space usage
+  - ```rust
+  struct SURI {
+    userinfo: Option<ParaId>,
+    host: Option<ChainId>,
+    path_components: Vec<RuntimeStr>,
+    query: Option<RuntimeStr>,
+    fragment: Option<RuntimeStr>
+  };
+  type RuntimeStr = Vec<u8>;
+  type ParaId = u32; // per Polkadot implementation
+  enum ChainId {
+    Id(u32), // numeric chain id registry is out of scope of this proposol
+    Name(RuntimeStr)
+  };
+  ```
+
+## Tests
+
+(using `polkadot` as the schema)
+
+- Polkadot DOT: `polkadot://polkadot/balances`
+- The Default instance Balances token on parachain 1: `polkadot://1@polkadot/srml-balances`
+- The Instance2 Balances token on parachain 2: `polkadot://2@polkadot/balances.2`
+- Generic Asset Asset ID 5 token on a independent chain: `polkadot://chain-name/srml-generic-asset/5`
+- The smart contract module on parachain 3 on Kusama: `polkadot://3@kusama/srml-contracts`
+- A NFT with ID of  `UniqueWeapon001` with provided by a smart contract with address of `0x1234abcd` on a Instance2 contracts module on parachain 4 on Kusama: `polkadot://4@kusama/srml-contracts.2/0x1234abcd/UniqueWeapon001`
+- The balances of user `0x1111aaaa` of a ERC20-like token contract with address of `0x2222bbbb` on a smart contract chain: `polkadot://smart-contract-chain-name/srml-contracts/0x2222bbbb/0x1111aaaa`
```

Comment:
Are these only for reads? What if I want a URI to _send_ a transaction? Would it make sense to go with something similar to EIP-681 for those formats and include them in this?

Reply:
It is definitely to make something similar to RESTful, but maybe later? I want to limit the scope of this to just an identifier / locator.